src/librustdoc/clean/inline.rs

```diff
@@ -444,24 +445,32 @@ pub(crate) fn build_impl(
 
     let associated_trait = tcx.impl_trait_ref(did).map(ty::EarlyBinder::skip_binder);
 
+    // Do not inline compiler-internal items unless we're a compiler-internal crate.
+    let is_compiler_internal = |did| {
+        if let Some(stab) = tcx.lookup_stability(did) {
+            stab.is_unstable() && stab.feature == sym::rustc_private
+        } else {
+            false
+        }
+    };
+    let document_compiler_internal = is_compiler_internal(LOCAL_CRATE.as_def_id());
+    let is_directly_public = |cx: &mut DocContext<'_>, did| {
+        if !cx.cache.effective_visibilities.is_directly_public(tcx, did) {
+            return false;
+        }
+        if !document_compiler_internal && is_compiler_internal(did) {
+            return false;
+        }
+        true
+    };
+
     // Only inline impl if the implemented trait is
     // reachable in rustdoc generated documentation
     if !did.is_local()
         && let Some(traitref) = associated_trait
+        && !is_directly_public(cx, traitref.def_id)
     {
-        let did = traitref.def_id;
-        if !cx.cache.effective_visibilities.is_directly_public(tcx, did) {
-            return;
-        }
-
-        if !tcx.features().rustc_private && !cx.render_options.force_unstable_if_unmarked {
```

Comment:
quick question, why wouldn't the following be a potential solution:

```rs
if !cx.render_options.force_unstable_if_unmarked {
```

(just removing the F-rustc_private check)